src/allocator_global_pool.hpp

```diff
@@ -0,0 +1,137 @@
+    copy and distribute the resulting executable under terms of your choice,
+    provided that you also meet, for each linked independent module, the
+    terms and conditions of the license of that module. An independent
+    module is a module which is not derived from or based on this library.
+    If you modify this library, you must extend this exception to your
+    version of the library.
+
+    libzmq is distributed in the hope that it will be useful, but WITHOUT
+    ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+    FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+    License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef __ZMQ_MEMORYPOOL_HPP_INCLUDED__
+#define __ZMQ_MEMORYPOOL_HPP_INCLUDED__
+
+#include "msg.hpp"
+#include <vector>
+
+#if (defined __cplusplus && __cplusplus >= 201103L)
+#include "../external/mpmcqueue/concurrentqueue.h"
+#else
+#include "basic_concurrent_queue.hpp"
+#endif
+#include "mutex.hpp"
+
+#define ZMQ_GLOBAL_POOL_FIRST_BLOCK_SIZE (256)
+
+namespace zmq
+{
+class allocator_global_pool_t
+{
+  public:
+    static void *allocate_fn (void *allocator_, size_t len_)
+    {
+        return static_cast<allocator_global_pool_t *> (allocator_)
+          ->allocate (len_);
+    }
+
+    static void deallocate_fn (void *allocator_, void *data_)
+    {
+        return static_cast<allocator_global_pool_t *> (allocator_)
+          ->deallocate (data_);
+    }
+
+    static bool check_tag_fn (void *allocator_)
+    {
+        return static_cast<allocator_global_pool_t *> (allocator_)
+          ->check_tag ();
+    }
+
+    allocator_global_pool_t (size_t initialMaximumBlockSize = 8192);
+    ~allocator_global_pool_t ();
+
+    void allocate_block (size_t bl);
+
+    // TODO have a look if realloc is possible, probably not as not thread safe as messages might still be in-flight?
+    void expand_block (size_t bl);
+
+    void *allocate (size_t len); // consumer thread: user app thread
+
+    void deallocate (void *data_); // producer thread: ZMQ background IO thread
+
+    size_t size () const;
+
+    bool check_tag () const;
+
+  private:
+    uint32_t _tag;
+
+    typedef struct
+    {
+        size_t num_msgs;
+        // actual user data
+        std::vector<uint8_t *> raw_data;
+    } msg_block_t;
+
+    std::vector<msg_block_t> _storage;
+#if (defined __cplusplus && __cplusplus >= 201103L)
+    std::vector<moodycamel::ConcurrentQueue<uint8_t *> > _free_list;
+#else
+    std::vector<basic_concurrent_queue_t<uint8_t *> > _free_list;
+#endif
+    mutex_t _storage_mutex;
+
+    inline size_t MsgBlockToBytes (size_t block)
+    {
+        return ZMQ_GLOBAL_POOL_FIRST_BLOCK_SIZE * 2 ^ block;
+    }
+
+    // by Todd Lehman https://stackoverflow.com/questions/994593/how-to-do-an-integer-log2-in-c
```

Comment:
IIRC all code snippets from SO are licensed under CC-BY-SA v3 which is not compatible with the GPL license

Reply:
Removed and replaced by somewhat less efficient (but more readable) implementation. Shouldn't impact performance too much.